File: cpp/banco.cpp
class usuario: 
    
    def __init__(self, nombre, pin, saldo):
        self.nombre = nombre
        self.pin = pin
        self.saldo = saldo

class Banco: 
     
    def __init__(self, usuarios=[]):
        self.usuarios=usuarios
    def autenticar(self, nombre, pin):
        for usuario in self.usuarios:
            if usuario.nombre == nombre:
                if usuario.pin == pin:
                    print ("Estas logueado")
                    return True
                else:
                    print("pin Incorrecto")
                    return False
        else:
            print ("El usuario no esxiste")
            return  False
    def sacar_dineo(self, nombre, saldo):
        for usuario in self.usuarios:
            if usuario.nombre == nombre:
                if usuario.saldo < saldo:
                    print("saldo insufiente")
                    break
                elif usuario.saldo >=saldo:
                    usuario.saldo -=saldo
                    print("saldo disponible es", usuario.saldo)
                    
ana = usuario ("Ana", 9872, 450)
pablo = usuario("Pablo", 5555, 200)
rodrigo = usuario("Rodrigo", 2222, 900) 

banco = Banco(usuarios=[ana, pablo,rodrigo])
while True:
    print("Bienvenido al banco. por favor, identifiquese.")
    print("introduzca su nombre")
    nombre = input()
    print("Introduzca el pin")
    pin = int(input())
    
    if banco.autenticar(nombre, pin):
        while True:
            print("por favor, elija de estas opciones.\n1. sacar dinero\n2.terminar sesión")
            opcion = input()
            
            if opcion=="1:":
                print("Introduzca cantidad a sacar")
                saldo = int(input())
                banco.sacar_dinero(nombre, saldo)
          
            elif opcion == "2":
                print("saliendo del sistema")
                break    
            else:
                print("opcion incorrecta. Por favor, introduzca otra opcion")
        else:
            print("usuario no autenticado, por favor, introduzca un nombre y pin correctos") 

File: java/banco.java
import java.util.ArrayList;
import java.util.Scanner;

class Usuario {
    String nombre;
    int pin;
    double saldo;

    public Usuario(String nombre, int pin, double saldo) {
        this.nombre = nombre;
        this.pin = pin;
        this.saldo = saldo;
    }
}

class Banco {
    ArrayList<Usuario> usuarios;

    public Banco(ArrayList<Usuario> usuarios) {
        this.usuarios = usuarios;
    }

    public boolean autenticar(String nombre, int pin) {
        for (Usuario usuario : usuarios) {
            if (usuario.nombre.equals(nombre)) {
                if (usuario.pin == pin) {
                    System.out.println("Estas logueado");
                    return true;
                } else {
                    System.out.println("Pin Incorrecto");
                    return false;
                }
            }
        }
        System.out.println("El usuario no existe");
        return false;
    }

    public void sacarDinero(String nombre, double monto) {
        for (Usuario usuario : usuarios) {
            if (usuario.nombre.equals(nombre)) {
                if (usuario.saldo < monto) {
                    System.out.println("Saldo insuficiente");
                    return;
                } else {
                    usuario.saldo -= monto;
                    System.out.println("Saldo disponible es: " + usuario.saldo);
                }
            }
        }
    }
}

public class Main {
    public static void main(String[] args) {
        // Crear usuarios
        Usuario ana = new Usuario("Ana", 9872, 450);
        Usuario pablo = new Usuario("Pablo", 5555, 200);
        Usuario rodrigo = new Usuario("Rodrigo", 2222, 900);

        // Crear una lista de usuarios y banco
        ArrayList<Usuario> usuarios = new ArrayList<>();
        usuarios.add(ana);
        usuarios.add(pablo);
        usuarios.add(rodrigo);

        Banco banco = new Banco(usuarios);

        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.println("Bienvenido al banco. Por favor, identifíquese.");
            System.out.print("Introduzca su nombre: ");
            String nombre = scanner.nextLine();
            System.out.print("Introduzca el pin: ");
            int pin = scanner.nextInt();
            scanner.nextLine(); // Limpiar buffer

            if (banco.autenticar(nombre, pin)) {
                while (true) {
                    System.out.println("Por favor, elija una opción:\n1. Sacar dinero\n2. Terminar sesión");
                    String opcion = scanner.nextLine();

                    if (opcion.equals("1")) {
                        System.out.print("Introduzca la cantidad a sacar: ");
                        double saldo = scanner.nextDouble();
                        scanner.nextLine(); // Limpiar buffer
                        banco.sacarDinero(nombre, saldo);

                    } else if (opcion.equals("2")) {
                        System.out.println("Saliendo del sistema");
                        break;

                    } else {
                        System.out.println("Opción incorrecta. Por favor, introduzca otra opción");
                    }
                }
            } else {
                System.out.println("Usuario no autenticado, por favor, introduzca un nombre y pin correctos.");
            }
        }
    }
}
